utils/python/CIME/XML/entry_id.py

```diff
@@ -157,12 +157,18 @@ def _set_value(self, node, value, vid=None, subgroup=None, ignore_type=False):
         subgroup is ignored in the general routine and applied in specific methods
         """
         expect(subgroup is None, "Subgroup not supported")
+        valid_values = self.get_optional_node("valid_values", root=node)
         if ignore_type:
             expect(type(value) is str, "Value must be type string if ignore_type is true")
-            node.set("value",value)
+            str_value = value
         else:
             type_str = self._get_type_info(node)
-            node.set("value", convert_to_string(value, type_str, vid))
+            str_value = convert_to_string(value, type_str, vid)
+        if valid_values is not None and valid_values.text is not None and not str_value.startswith('$'):
```

Comment:
I think we should close this PR and create an issue for the points @gold2718 raised. The code is ok and does what is required.